packages/@lwc/template-compiler/src/shared/ir.ts

```diff
@@ -4,48 +4,82 @@
  * SPDX-License-Identifier: MIT
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
  */
+import * as parse5 from 'parse5';
+import treeAdapter from 'parse5/lib/tree-adapters/default';
+
 import {
     TemplateIdentifier,
     TemplateExpression,
     IRNode,
     IRText,
     IRElement,
-    HTMLElement,
-    HTMLText,
-    HTMLComment,
     IRComment,
 } from './types';
 
-export function createElement(original: HTMLElement, parent?: IRElement): IRElement {
+export function createElement(original: parse5.Element, parent?: IRElement): IRElement {
+    let location = original.sourceCodeLocation;
+
+    // With parse5 automatically recovering from invalid HTML, some AST nodes might not have
+    // location information. For example when a <table> element has a <tr> child element, parse5
+    // creates a <tbody> element in the middle without location information. In this case, we
+    // can safely skip the closing tag validation.
+    //
+    // TODO [#248]: Report a warning when location is not available indicating the original HTML
+    // template is not valid.
+    let current = original;
+    while (!location && treeAdapter.isElementNode(original.parentNode)) {
```

Comment:
```suggestion
    while (!location && original.parentNode.nodeName !== '#document') {
```
Making it explicit that we stop at the root.